packages/deployments/contracts/contracts/SponsorVault.sol

```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.11;
+
+import {ISponsorVault} from "./interfaces/ISponsorVault.sol";
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20, Address} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+interface ITokenExchange {
+  function getInGivenExactOut(address token, uint256 exactOut) external returns (uint256);
+  function swapExactOut(address token, uint256 exactOut, address recipient) external payable returns (uint256);
+}
+
+interface IGasTokenOracle {
+  function getRate(uint32 originDomain) external view returns (uint256 num, uint256 den);
+}
+
+/**
+ * @title SponsorVault
+ * @author Connext Labs
+ * @notice Contains logic for sponsoring liquidity and relayer fees
+ */
+contract SponsorVault is ISponsorVault, Ownable{
+  // ============ Libraries ============
+  using SafeERC20 for IERC20;
+
+  // ============ Struct ============
+  struct Rate {
+    uint256 num;
+    uint256 den;
+  }
+
+  // ============ Private storage ============
+
+  // ============ Public storage ============
+
+  /**
+   * @notice The address of connext
+   */
+  address public connext;
+
+  /**
+   * @notice The origin domain to this domain gas token rates
+   * @dev Used when no oracle is available
+   */
+  mapping(uint32 => Rate) public rates;
+
+  /**
+   * @notice The maximum amount of relayer fee to sponsor
+   */
+  uint256 public relayerFeeCap;
+
+  /**
+   * @notice The origin domain to this domain gas token oracle
+   * @dev Used to calculate sponsored relayer fee
+   */
+  address public gasTokenOracle;
+
+  /**
+   * @notice The this domain gas token to token exchange
+   * @dev Used to calculate sponsored relayer fee
+   */
+  mapping(address => address payable) public tokenExchanges;
+
+  // ============ Errors ============
+
+  error SponsorVault__setConnext_invalidConnext();
+  error SponsorVault__setRate_invalidOriginDomain();
+  error SponsorVault__setGasTokenOracle_invalidOriginDomain();
+  error SponsorVault__setTokenExchange_invalidAdopted();
+  error SponsorVault__onlyConnext();
+
+  // ============ Events ============
+
+  /**
+   * @notice Emitted when a new connext is set
+   */
+  event ConnextUpdated(address connext, address caller);
+
+  /**
+   * @notice Emitted when a new rate is set
+   */
+  event RateUpdated(uint32 originDomain, Rate rate, address caller);
+
```

Comment:
I am wondering why we have `caller` address on all events. Actually, we can get `caller` from `transaction.from`  on subgraph handlers. If there is no special reason, it will be inefficient.

Reply:
I followed the pattern from the events defined in Connext.
Regarding getting que `caller` from the `transaction.from` on the subgraph, I think it would have the address of the transaction origin so if the this contract is called from another contract (some governance contract maybe) `caller` is the address of the caller contract  but `transaction.from` is the address of the EOA which initiates the transaction.

Reply:
okay. then that is fine